### PR TITLE
Add Authentication and Authorization logic.

### PR DESCRIPTION
**Pull request by AndreiDrang**

### CODE REVIEW

**Review by kurkl — APPROVED** [review]

File: spectree/models.py [hunk 1]
```diff
@@ -44,3 +45,19 @@ class UnprocessableEntity(BaseModel):
...
+
+
+class SecureType(str, Enum):
+    HTTP = "http"
+    API_KEY = "apiKey"
+    OAUTH_TWO = "oauth2"
+    OPEN_ID_CONNECT = "openIdConnect"
+
+
+class SecuritySchemesData(BaseModel):
```
Comment (kemingy) [inline spectree/models.py:57]: Can export this to `spectree/__init__.py` file for users to import.

File: spectree/config.py [hunk 2]
```diff
@@ -63,3 +68,6 @@ def update(self, **kwargs):
 
         assert self.UI in self._SUPPORT_UI, "unsupported UI"
         assert self.MODE in self._SUPPORT_MODE, "unsupported MODE"
+        if self.AUTH_METHODS:
+            for value in self.AUTH_METHODS.values():
+                SecuritySchemesData(**value)
```
Comment (kemingy) [inline spectree/config.py:73]: Will it be better to accept the instance of `SecuritySchemesData`? So when the user creates this, the data is already validated. Then we can add another method to get the JSON schema for OpenAPI to use.

**Review by kemingy — COMMENTED** [review]
Awesome! Really appreciate your contribution.

Another thing I want to talk about is, for Authentication Error, should we return ValidationError(422) or AuthenticationError(401 or 403)?

File: spectree/config.py (hunk 2 above)
Comment (AndreiDrang) [reply spectree/config.py:73]: Also, I'm not sure about `self.AUTH_METHODS` naming, maybe name from the OpenAPI scheme will be better?
For example - `self.securitySchemes`, not a pythonic style but... it is easy to understand what the parameter it is.

Comment (kemingy) [reply spectree/config.py:73]: Agree. `self.SECURITY_SCHEMES` sounds better. BTW, I'm thinking if we should use `pydantic.BaseSettings` for the config class.

Comment (AndreiDrang) [reply spectree/config.py]: > pydantic.BaseSettings

Sounds good but for some new PR.

**Comment by AndreiDrang** [comment]
> Another thing I want to talk about is, for Authentication Error, should we return ValidationError(422) or AuthenticationError(401 or 403)?

I'm not sure that hardcoded Errors are good, maybe some flexibility is better.

Cause, for example in my case I not use 422 error from swagger config, I always return 200 code and put error info in payload under specific key, and frontend developers ask me questions about "422 error from swagger" and when it rises :smiley:

But I think, that `401` error will be better. Or we can just print some `WARNING` message(like SQLAlchemy) if user sets `security` key for route, but not set `401` or `403` error in resp.
image

File: spectree/models.py [hunk 3]
```diff
@@ -1,7 +1,11 @@
+import re
 from enum import Enum
 from typing import Any, Dict, Sequence
 
-from pydantic import BaseModel, Field
+from pydantic import BaseModel, Field, validator
+
+# OpenAPI names validation regexp
+OAPI_NAME_RE = re.compile(r"^[A-Za-z0-9-._]+")
```
Comment (kemingy) [inline spectree/models.py:8]: Can choose one from (`OpenAPI_NAME_RE`, `OPENAPI_NAME_RE`). `OAPI_NAME_RE` is a little bit confusing.

File: spectree/models.py [hunk 4]
```diff
@@ -54,10 +58,38 @@ class SecureType(str, Enum):
...
+
+class SecurityScheme(BaseModel):
+    """
+    Named security scheme
+    """
+
+    name: str = Field(
+        ...,
+        description="Custom security scheme name. Can only contain - [A-Za-z0-9-._]",
+    )
```
Comment (kemingy) [inline spectree/models.py:85]: What's the difference between `SecurityScheme.name` and `SecuritySchemeData.name`?

File: spectree/models.py [hunk 5]
```diff
@@ -54,10 +58,38 @@ class SecureType(str, Enum):
...
+    """
+    Security scheme data
+    """
+
     type: SecureType = Field(..., description="Secure scheme type")
     name: str = None
     field_in: str = Field(None, alias="in")
     scheme: str = None
     openIdConnectUrl: str = None
     flows: dict = None
```
Comment (kemingy) [inline spectree/models.py]: Except `type`, all the other items are optional. So if the user create `SecuritySchemeData(type="http")`, should that be accepted?

**Comment by kemingy** [comment]
> I'm not sure that hardcoded Errors are good, maybe some flexibility is better.
> 
> Cause, for example in my case I not use 422 error from swagger config, I always return 200 code and put error info in payload under specific key, and frontend developers ask me questions about "422 error from swagger" and when it rises 😃
> 
> But I think, that `401` error will be better. Or we can just print some `WARNING` message(like SQLAlchemy) if user sets `security` key for route, but not set `401` or `403` error in resp.

We can use `401` as the default one. I also provide `before_hook` for the user to change the request validation error response body.

File: spectree/models.py (hunk 4 above)
Comment (AndreiDrang) [reply spectree/models.py:85]: > SecurityScheme.name

Custom name for auth scheme, init in `components->securitySchemes` and then can be used in `route->method->security`. Need validation.

> SecuritySchemeData.name

Custom variable name for header, for example. Validation not needed.

Screenshot_2021-05-05_21-39-58

Comment (AndreiDrang) [reply spectree/models.py:130]: Add more descriptions

File: spectree/models.py (hunk 5 above)
Comment (AndreiDrang) [reply spectree/models.py]: Fix

**Comment by kemingy** [comment]
Please run `make lint` and `make test` before the push.

**Comment by Qparrot** [comment]
Hello, 
Do you know if it is now possible to add Authorization logic ? 
When I search for "security_scheme", "authorization" I found no related page on the [SpecTree documentation](https://spectree.readthedocs.io/en/latest/).

thank you for all the work :), Spectree is great!

**Comment by kemingy** [comment]
> Do you know if it is now possible to add Authorization logic ?

@Qparrot I haven't added docs for this part. But you can refer to the test file https://github.com/0b01001001/spectree/blob/f5e56e57d580cdc78d67af20737c0a346246d595/tests/test_plugin_flask.py#L38 and the examples on the bottom of that file.

**Comment by kemingy** [comment]
> Hello,
> Do you know if it is now possible to add Authorization logic ?
> When I search for "security_scheme", "authorization" I found no related page on the [SpecTree documentation](https://spectree.readthedocs.io/en/latest/).

The new PR #151 added the related document in the README. @Qparrot

**Pull request merged**